classes/user/Collector.inc.php

```diff
@@ -394,183 +408,159 @@ public function getQueryBuilder(): Builder
-                        return $join->on("${alias}.user_id", '=', 'u.user_id')
-                            ->where("${alias}.setting_name", '=', Identity::IDENTITY_SETTING_GIVENNAME)
-                            ->where("${alias}.locale", '=', $locale);
-                    });
+            ->when($this->offset !== null, function (Builder $query): void {
+                if ($this->count === null) {
+                    throw new InvalidArgumentException('The offset requires the count to be defined');
                 }
-                $q->addSelect([DB::raw('COALESCE(' . implode('.setting_value, ', $aliases) . '.setting_value) AS given_name')])
-                    ->orderBy('given_name', $this->orderDirection);
-                break;
-            case self::ORDERBY_FAMILYNAME:
-                $aliases = [];
-                foreach ($orderLocales as $key => $locale) {
-                    $alias = $aliases[] = "usf{$key}";
-                    $q->leftJoin("user_settings AS ${alias}", function ($join) use ($alias, $locale) {
-                        return $join->on("${alias}.user_id", '=', 'u.user_id')
-                            ->where("${alias}.setting_name", '=', Identity::IDENTITY_SETTING_FAMILYNAME)
-                            ->where("${alias}.locale", '=', $locale);
-                    });
+                $query->offset($this->offset);
+            })
+
+            ->when(
+                in_array($this->orderBy, [self::ORDERBY_GIVENNAME, self::ORDERBY_FAMILYNAME]),
+                function (Builder $query): void {
+                    $query->orderBy(function (Builder $query): void {
+                        $settingMap = [self::ORDERBY_GIVENNAME => Identity::IDENTITY_SETTING_GIVENNAME, self::ORDERBY_FAMILYNAME => Identity::IDENTITY_SETTING_FAMILYNAME];
+                        $locales = array_unique(empty($this->orderLocales) ? [AppLocale::getLocale(), Application::get()->getRequest()->getSite()->getPrimaryLocale()] : $this->orderLocales);
+                        $lastIndex = 0;
+                        $bindinds = array_reduce($locales, function (array $bindings, string $locale) use (&$lastIndex): array {
+                            array_push($bindings, $locale, ++$lastIndex);
+                            return $bindings;
+                        }, []);
+                        $bindinds[] = ++$lastIndex;
+                        
+                        $query->from('user_settings', 'us')
+                            ->whereColumn('us.user_id', '=', 'u.user_id')
+                            ->where('us.setting_name', '=', $settingMap[$this->orderBy])
+                            ->whereIn('us.locale', $locales)
+                            ->orderByRaw("COALESCE(us.setting_value, '') = ''") // Prefers non-null/non-empty values
+                            ->orderByRaw('CASE us.locale ' . str_repeat('WHEN ? THEN ?', count($locales)) . ' ELSE ? END', [$bindinds]) // Then follow the locale order
```

Comment:
Should `$bindinds` be supplied as wrapped in an array? It looks to me like it should be passed in as-is. But I'm finding this segment of code a little hard to read, so I could be wrong.

Reply:
It works this way as well (there's a `array_merge_recursive` in the internals), so I didn't notice =]